### PR TITLE
Add method: transfernft & show nft

### PR DESCRIPTION
**Pull request by chenzhitong**
close https://github.com/neo-project/neo-express/issues/193 

**transfernft**
```
Usage: neoxp transfernft [options] <Contract> <TokenId> <Sender> <Receiver>

Arguments:
  Contract                  NFT Contract (symbol or script hash)
  TokenId                   TokenId of NFT (Base64 string)
  Sender                    Account to send NFT from
  Receiver                  Account to send NFT to

Options:
  -d|--data <DATA>          Optional data parameter to pass to transfer operation
  -p|--password <PASSWORD>  password to use for NEP-2/NEP-6 sender
  -i|--input <INPUT>        Path to neo-express data file
  -t|--trace                Enable contract execution tracing
  -j|--json                 Output as JSON
  -?|-h|--help              Show help information.
```
image

**show nft**

```
Usage: neoxp show nft [options] <Contract> <Account>

Arguments:
  Contract            Contract to show NFT of (symbol or script hash)
  Account             Account to show asset balance for

Options:
  -i|--input <INPUT>  Path to neo-express data file
  -?|-h|--help        Show help information.
```
image

### CODE REVIEW

File: src/neoxp/Commands/TransferNFTCommand.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        internal bool Json { get; init; } = false;
+
+        internal async Task<int> OnExecuteAsync(CommandLineApplication app, IConsole console)
+        {
+            try
+            {
+                var (chainManager, _) = chainManagerFactory.LoadChain(Input);
+                var password = chainManager.Chain.ResolvePassword(Sender, Password);
+                using var txExec = txExecutorFactory.Create(chainManager, Trace, Json);
+                await txExec.TransferNFTAsync(Contract, Encoding.UTF8.GetString(Convert.FromBase64String(TokenId)), Sender, password, Receiver, Data).ConfigureAwait(false);
```
Comment (cschuchardt88) [inline src/neoxp/Commands/TransferNFTCommand.cs:57]: TokenId Should be in `hex`. This is the way it is done in production. Will be easier to understand.

File: src/neoxp/Commands/TransferNFTCommand.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+        [Argument(3, Description = "Account to send NFT to")]
+        [Required]
+        internal string Receiver { get; init; } = string.Empty;
+
+        [Option(Description = "Optional data parameter to pass to transfer operation")]
+        internal string Data { get; init; } = string.Empty;
+
+        [Option(Description = "password to use for NEP-2/NEP-6 sender")]
+        internal string Password { get; init; } = string.Empty;
```
Comment (cschuchardt88) [inline src/neoxp/Commands/TransferNFTCommand.cs:39]: `NEP-17` transfers dont require password. Seeing how there is `default.neo-express` file that holds it.

File: src/neoxp/neoxp.csproj [hunk 3]
```diff
@@ -12,6 +12,10 @@
     <PackageReference Include="Nito.Disposables" Version="2.3.0" />
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Update="Neo.BlockchainToolkit.Library" Version="3.6.19" />
+  </ItemGroup>
+
```
Comment (cschuchardt88) [inline src/neoxp/neoxp.csproj:18]: `3.6.19` doesn't exists anymore its `3.6.0` now. But we have updated repo with lots of changed please merge before adding this. If still needed then keep it.

File: src/neoxp/Commands/TransferNFTCommand.cs (hunk 2 above)
Comment (chenzhitong) [reply src/neoxp/Commands/TransferNFTCommand.cs:39]: This is copied from TransferCommand.cs and is the same as NEP-17.

File: src/neoxp/Properties/launchSettings.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "profiles": {
+    "neoxp": {
+      "commandName": "Project",
+      "commandLineArgs": "run"
+    }
+  }
+}
```
Comment (cschuchardt88) [inline src/neoxp/Properties/launchSettings.json:8]: Please dont merge your `launchSettings.json`.

**Review by cschuchardt88 — COMMENTED** [review]
Dont forget to run `dotnet format` and your analyzers to fix github actions.

File: src/neoxp/Commands/TransferNFTCommand.cs (hunk 2 above)
Comment (cschuchardt88) [reply src/neoxp/Commands/TransferNFTCommand.cs:39]: yes but for `password to use for NEP-2/NEP-6 sender` for legacy i believe.

File: src/neoxp/Commands/TransferNFTCommand.cs [hunk 5]
```diff
@@ -54,14 +56,29 @@ internal async Task<int> OnExecuteAsync(CommandLineApplication app, IConsole con
...
+                hex = hex.ToLower().Trim();
+                if (!new Regex("^(0x)?([0-9a-f]{2})+$").IsMatch(hex)) throw new FormatException();
+
+                if (new Regex("^([0-9a-f]{2})+$").IsMatch(hex))
+                {
+                    return Encoding.UTF8.GetString(hex.HexToBytes());
+                }
+                else
+                {
+                    return Encoding.UTF8.GetString(hex[2..].HexToBytes().Reverse().ToArray());
```
Comment (cschuchardt88) [inline src/neoxp/Commands/TransferNFTCommand.cs:79]: `Convert.FromHexString` or `Convert.ToHexString` is good too.

File: src/neoxp/Commands/TransferNFTCommand.cs (hunk 2 above)
Comment (cschuchardt88) [reply src/neoxp/Commands/TransferNFTCommand.cs:41]: Also keep in mind for `batch` it doesn't require password for transfers. That's what i was thinking of.

**Comment by cschuchardt88** [comment]
since you are  working on #193 can you also checkout and think adding `NEP-11` for #194

**Comment by cschuchardt88** [comment]
You have forgotten about Divisible NFTs

Please see
https://github.com/neo-project/proposals/blob/master/nep-11.mediawiki#user-content-transfer-2

.vs
https://github.com/neo-project/proposals/blob/master/nep-11.mediawiki#user-content-transfer

File: src/neoxp/Extensions/ExpressNodeExtensions.cs [hunk 6]
```diff
@@ -396,6 +405,36 @@ public static async Task<(RpcNep17Balance balance, Nep17Contract token)> GetBala
...
+                {
+                    while (iterator.Next())
+                    {
+                        list.Add(iterator.Value(null).GetString());
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                throw new Exception("invalid script results");
```
Comment (cschuchardt88) [inline src/neoxp/Extensions/ExpressNodeExtensions.cs:433]: Can we get a better error output? Something indicating that the contract isn't an NFT or missing a method

**Comment by chenzhitong** [comment]
> You have forgotten about Divisible vs Non-Divisible NFTs
> 
> Please see https://github.com/neo-project/proposals/blob/master/nep-11.mediawiki#user-content-transfer-2
> 
> .vs https://github.com/neo-project/proposals/blob/master/nep-11.mediawiki#user-content-transfer

Template for [neo-devpack-dotnet](https://github.com/neo-project/neo-devpack-dotnet/blob/master/src/Neo.SmartContract.Framework/Nep11Token.cs)  only has base class for Non-divisible NFT, so, I think just supporting Non-divisible NFT is enough. 

BTW, does any wallet support Divisible NFT now?

**Comment by cschuchardt88** [comment]
>BTW, does any wallet support Divisible NFT now?

image
I believe NeoLine. That what the Number `1` is for? But I don't think we should limit it.

>Template for [neo-devpack-dotnet](https://github.com/neo-project/neo-devpack-dotnet/blob/master/src/Neo.SmartContract.Framework/Nep11Token.cs) only has base class for Non-divisible NFT, so, I think just supporting Non-divisible NFT is enough.

Once we add the validator for NEP-11 this won't be a problem. That is because in `neo-devpack-dotnet` if you want divisible you just have to add these functions `transfer` `ownerOf` and `balanceOf`. Also seeing how the class is abstract the non-divisible once won't be on the interface.

**Comment by chenzhitong** [comment]
image
This count is to the number of NFTs held in the same contract
When you click on it, you will see the details

image

**Comment by chenzhitong** [comment]
Who can review it

**Comment by Liaojinghui** [comment]
Lacks unit test, all i can do is merge when you think its good to go.

**Comment by chenzhitong** [comment]
I've tested it. Merge it please. @Liaojinghui

**Review by Liaojinghui — APPROVED** [review]

**Comment by Liaojinghui** [comment]
@chenzhitong @cschuchardt88 @Ashuaidehao Can you add unit test to this repo? I remember that @cschuchardt88  you wanted to add something that contains the unit test to this repo, right? I think we can just make things easy.

**Comment by cschuchardt88** [comment]
>@chenzhitong @cschuchardt88 @Ashuaidehao Can you add unit test to this repo? I remember that @cschuchardt88 you wanted to add something that contains the unit test to this repo, right? I think we can just make things easy.

I wanted to add `neo-blockchaintoolkit-library` to this repo that's where all the unit tests are for core functionally.


@chenzhitong 
can you make sure you put file header in `*.cs` files please.

In Visual Studio set it up like this
image

Than you can run like so if needed:
image

**Comment by cschuchardt88** [comment]
@chenzhitong can you reopen and fix?

**Comment by cschuchardt88** [comment]
- [ ] Add TokenIds as base64 as well
- [ ] Fix error output to show inter error
- [ ] Make wording say `[Argument(1, Description = "TokenId of NFT (Format: HEX, BASE64)")]`
- [ ]  Make wording say `[Argument(0, Description = "NFT Contract (Symbol or Script Hash)")]`

Try not to use Regex makes it hard to quickly understand. Why you use `Reverse` when you have `0x`?

**Pull request merged**